File: tiq-e/linked-list/3/main.go
package main

import (
	"fmt"
	"time"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseList(head *ListNode) *ListNode {
	var tail *ListNode
	var helper func(*ListNode, *ListNode)
	helper = func(prev, curr *ListNode) {
		if curr == nil {
			tail = prev
			return
		}

		helper(curr, curr.Next)
		curr.Next = prev
	}
	helper(nil, head)
	return tail
}

func main() {
	var head = ListNode{Val: 1, Next: nil}

	var start = time.Now()
	var result = reverseList(&head)
	var end = time.Since(start)
	fmt.Println("Duration: ", end)
	fmt.Println("Result: ", result)
}

File: tiq-e/linked-list/3/main.rs
use std::time::Instant;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

struct Solution;

impl Solution {
    pub fn reverse_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut cur = head;
        let mut prev: Option<Box<ListNode>> = None;

        while let Some(c) = cur {
            prev = Some(Box::new(ListNode {
                val: c.val,
                next: prev,
            }));
            cur = c.next;
        }

        prev
    }
}

fn main() {
    let head = Option::from(Box::new(ListNode { val: 1, next: None }));

    let start = Instant::now();
    let result = Solution::reverse_list(head);
    let end = start.elapsed().as_micros();
    println!("Duration: {:?}", end);
    println!("Result: {:?}", result);
}
